aerospike/check.py

```diff
@@ -3,75 +3,173 @@
 
         if conn is None:
             conn = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-            conn.connect((host, port))
-            self.connection_pool[key] = conn
+            conn.connect(addr)
+            self.connections[addr] = conn
 
-        try:
-            with closing(conn.makefile('r')) as fp:
-                if namespaces is None:
-                    # Probe all namespaces by default.
-                    conn.send('namespaces\r\n')
-                    namespaces = fp.readline().rstrip().split(';')
-
-                conn.send('statistics\r\n')
-                self._send_metrics(fp, want_metrics)
-
-                for n in namespaces:
-                    conn.send('namespace/%s\r\n' % n)
-                    tags = {'namespace': n}
-                    self._send_metrics(fp, want_namespace_metrics, tags)
-
-            self.service_check(SERVICE_CHECK_NAME, AgentCheck.OK)
-        except socket.error:
-            self.log.exception('Error connecting to Aerospike at %s', key)
-            del self.connection_pool[key]
-
-    def _send_metrics(self, fp, only_keys=[], tags={}):
-        vals = dict(x.split('=', 1) for x in fp.readline().rstrip().split(';'))
-        if only_keys:
-            # Cherry pick only desired metrics, if they exist.
-            for skey in only_keys:
-                value = vals.get(skey, None)
-                if value and value.isdigit():
-                    self.gauge(self._make_key(skey), value, tags=tags)
+        return conn
+
+    def _process_throughput(self, data, event_type, namespaces, tags={}):
+        while data != []:
+            l = data.pop(0)
+
+            # skip errors
+            while l.startswith('error-'):
+                if data == []:
+                    return
+                l = data.pop(0)
+
+            match = re.match('^{(.+)}-([^:]+):',l)
+            if match == None:
+                continue
+
+            ns = match.groups()[0]
+            if ns not in namespaces:
+                continue
+
+            key = match.groups()[1]
+            if data == []:
+                return # unexpected EOF
+            val = data.pop(0).split(',')[1]
+
+            self._send(event_type, key, val, tags + ['namespace:%s' % ns] )
+
+    def _process_data(self, fp, event_type, required_keys=[], tags={}):
+        d = dict(x.split('=', 1) for x in fp.readline().rstrip().split(';'))
+        if required_keys:
+            required_data = {k: d[k] for k in required_keys if k in d}
+        else:
+            required_data = d
+
+        for key, value in required_data.items():
+            self._send(event_type, key, value, tags)
+
+    def _send(self, event_type, key, val, tags={}):
+        datatype = 'event'
+
+        if re.match('^{(.+)}-^[-]+-hist',key):
+            return # skip histogram configuration
+
+        if key == 'cluster_key':
+            val = str(int(val, 16))
+
+        if val.isdigit():
```

Comment:
Is there any scenario where we might want to submit a `rate` instead of a `gauge`? I know that's not possible to detect in a generic way. But perhaps it's something we could add to the YAML as an attribute for a metric somehow (if you see a use for it). A `rate` would typically be useful for metrics that will always be increasing (for instance _number of requests since startup_, or whatever).

Reply:
Quiet a few, we provide a mapping in our collectd plugin:
https://github.com/aerospike/aerospike-collectd/blob/develop/aerospike_schema.yaml

The "rate" type would be classified under "operations" in our schema file.